File: java/src/main/java/com/ncc/aif/ClaimComponent.java
package com.ncc.aif;

import java.util.HashSet;
import java.util.Set;

import org.apache.jena.rdf.model.Model;
import org.apache.jena.rdf.model.Resource;

public class ClaimComponent {
    private String name;
    private String identity;
    private Set<String> types = new HashSet<>();
    private String provenance;
    private Resource ke;
    private Resource resource;

    public ClaimComponent setName(String name) {
        this.name = name;
        return this;
    }

    public ClaimComponent setIdentity(String identity) {
        this.identity = identity;
        return this;
    }

    public ClaimComponent setTypes(Set<String> types) {
        this.types = types;
        return this;
    }

    public ClaimComponent setProvenance(String provenance) {
        this.provenance = provenance;
        return this;
    }

    public ClaimComponent setKE(Resource ke) {
        this.ke = ke;
        return this;
    }

    public ClaimComponent addType(String type) {
        if (types == null) {
            types = new HashSet<>();
        }
        types.add(type);
        return this;
    }

    public String getName() {
        return this.name;
    }

    public String getIdentity() {
        return this.identity;
    }

    public Set<String> getTypes() {
        return this.types;
    }

    public String getProvenance() {
        return this.provenance;
    }

    public Resource getKE() {
        return this.ke;
    }

    public Resource getResource() {
        return this.resource;
    }

    public Resource addToModel(Model model, String uri, Resource system) {
        resource = AIFUtils.makeAIFResource(model, uri, InterchangeOntology.ClaimComponent, system)
            .addProperty(InterchangeOntology.componentName, name)
            .addProperty(InterchangeOntology.componentIdentity, identity);
        if (provenance != null) {
            resource.addProperty(InterchangeOntology.componentProvenance, provenance);
        }
        if (ke != null) {
            resource.addProperty(InterchangeOntology.componentKE, ke);
        }
        AIFUtils.addProperties(resource, InterchangeOntology.componentType, types);
        return resource;
    }
}

File: python/aida_interchange/claim_component.py
from enum import Enum

from rdflib import XSD, Literal, URIRef

from aida_interchange import aifutils
from aida_interchange.rdf_ontologies import interchange_ontology


class ClaimComponent:
    """
    This class encapsulates the LDC representation of time
    """

    def __init__(self):
        self._name          = None
        self._identity      = None
        self._type          = set(())
        self._provenance    = None
        self._ke            = None

    @property
    def name(self):
        return self._name
    @name.setter
    def setName(self, x):
        self._name = x
        
    @property
    def identity(self):
        return self._identity
    @identity.setter
    def setIdentity(self, x):
        self._identity = x

    @property
    def type(self):
        return self._type
    @type.setter
    def addType(self, x):
        self._type.add(x)

    @property
    def provenance(self):
        return self._provenance
    @provenance.setter
    def setProvenance(self, x):
        self._provenance = x

    @property
    def ke(self):
        return self._ke
    @ke.setter
    def setKe(self, x):
        self._ke = x


    def add_literal(self, g, claim_component, property, value, literal_type):
        """
        Creates and adds a Literal to the model if a value exists

        :param ClaimComponent self
        :param rdflib.graph.Graph g: The underlying RDF model for the operation
        :param rdflib.term.BNode claim_component: The resource created to contain claim component information
        :param time_property: AIDA_ANNOTATION.year, AIDA_ANNOTATION.month, or AIDA_ANNOTATION.day
        :param str value: the string value
        :param literal_type: the datatype that corresponds to the given value
        """
        if value is not None:
            claim_literal = Literal(value)
            claim_literal._datatype = URIRef(literal_type)
            g.add((claim_component, property,claim_literal))


    def make_aif_claim_component(self, g, claim_component):
        """
        Creates a claim component
        
        :param ClaimComponent self
        :param rdflib.graph.Graph g: The underlying RDF model for the operation
        :param claim_component: The parent RDF model for the operation
        :returns: node
        """
        
        self.add_literal(g, claim_component, interchange_ontology.componentName, self.name, XSD.string)
        
        self.add_literal(g, claim_component, interchange_ontology.componentIdentity, self.identity, XSD.string)

        for item in self.type:
            self.add_literal(g, claim_component, interchange_ontology.componentType, item, XSD.string)
        
        if (self.provenance != None):
            self.add_literal(g, claim_component, interchange_ontology.componentProvenance, self.provenance, XSD.string)

        if (self.ke != None):
            g.add((claim_component, interchange_ontology.componentKE, self.ke))  

        return claim_component
